traffic_ops/traffic_ops_golang/riak.go

```diff
@@ -0,0 +1,204 @@
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err := cluster.Stop(); err != nil {
+			log.Errorf("%v\n", err)
+		}
+	}()
+
+	if err = cluster.Start(); err != nil {
+		return err
+	}
+
+	// build store command and execute.
+	cmd, err := riak.NewStoreValueCommandBuilder().
+		WithBucket(bucket).
+		WithContent(obj).
+		Build()
+	if err != nil {
+		return err
+	}
+	if err := cluster.Execute(cmd); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// fetch an object from riak storage
+func fetchObject(key string, bucket string, db *sqlx.DB, cfg Config) (*riak.FetchValueCommand, error) {
+	// build the fetch command
+	cmd, err := riak.NewFetchValueCommandBuilder().
+		WithBucket(bucket).
+		WithKey(key).
+		Build()
+	if err != nil {
+		return nil, err
+	}
+	// create and start a riak cluster
+	cluster, err := getRiakCluster(db, cfg, 12)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err := cluster.Stop(); err != nil {
+			log.Errorf("%v\n", err)
+		}
+	}()
+	if err = cluster.Start(); err != nil {
+		return nil, err
+	}
+	if err = cluster.Execute(cmd); err != nil {
+		return nil, err
+	}
+	fvc := cmd.(*riak.FetchValueCommand)
+
+	return fvc, err
+}
+
+// endpoint handler for fetching uri signing keys from riak
+func urisignkeysHandler(db *sqlx.DB, cfg Config) RegexHandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request, p PathParams) {
+		handleErr := func(err error, status int) {
+			log.Errorf("%v %v\n", r.RemoteAddr, err)
+			w.WriteHeader(status)
+			fmt.Fprintf(w, http.StatusText(status))
+		}
+
+		xmlId := p["xml-id"]
+
+		fvc, err := fetchObject(xmlId, cdn_uri_keys_bucket, db, cfg)
+		if err != nil {
+			handleErr(err, http.StatusInternalServerError)
+			return
+		}
+
+		resp, err := getStringValue(fvc.Response)
+		if err != nil {
+			handleErr(err, http.StatusNotFound)
+			return
+		}
+		w.Header().Set("Content-Type", "application/json")
+		fmt.Fprintf(w, "%s", resp)
+	}
+}
+
+// returns a riak cluster of online riak nodes.
+func getRiakCluster(db *sqlx.DB, cfg Config, maxNodes int) (*riak.Cluster, error) {
+	riakServerQuery := fmt.Sprintf(`
+		SELECT s.host_name, s.domain_name FROM server s 
+		INNER JOIN type t on s.type = t.id 
+		INNER JOIN status st on s.status = st.id 
+		WHERE t.name = 'RIAK' AND st.name = 'ONLINE' LIMIT %d
+		`, maxNodes)
+
+	nodes := make([]*riak.Node, maxNodes)
+	rows, err := db.Query(riakServerQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	i := 0
+	count := 1
+	for rows.Next() {
+		var s tostructs.Server
+		if err := rows.Scan(&s.HostName, &s.DomainName); err != nil {
+			return nil, err
+		}
+		addr := fmt.Sprintf("%s.%s:%d", s.HostName, s.DomainName, riak_port)
+		nodeOpts := &riak.NodeOptions{
+			RemoteAddress: addr,
+			AuthOptions:   cfg.RiakAuthOptions,
+		}
+		nodes[i], err = riak.NewNode(nodeOpts)
+		if err != nil {
+			return nil, err
+		}
+		count++
+		i++
+	}
+
+	opts := &riak.ClusterOptions{
+		Nodes: nodes[0 : count-1],
```

Comment:
It appears `maxNodes` isn't exceptionally large. (Hardcoded to 12?) This subslice doesn't free the extra nodes you allocated, but didn't use. If, like it appears, this isn't a relevant amount of memory and isn't kept for an inordinate amount of time, that's totally fine. But if I'm misreading it, this could cause more memory to be used than you might guess. (The "fix" is to copy the memory, which might be more expensive, so I think this is still probably best.)